### PR TITLE
CL: fix CalcOutAmtGivenIn and other logic

### PR DESCRIPTION
**Pull request by czarcas7ic**
## What is the purpose of the change

This corrects faulty logic throughout the CL module, mostly within CalcOut and ComputeSwapStep

I think for now we should leave these tests as is, and have a subsequent issue convert it to a clean table driven test

This confirms OneToZero logic for CalcOut is good, I still need to check if ZeroToOne logic is sound (there is no reason why it shouldn't, but I didn't have time to check and wanted to get this merged to main ASAP)

### CODE REVIEW

File: x/concentrated-liquidity/math.go [hunk 1]
```diff
@@ -33,48 +33,59 @@ func liquidity1(amount sdk.Int, sqrtPriceA, sqrtPriceB sdk.Dec) sdk.Dec {
 // sqrtPriceA is the smaller of sqrtpCur and the nextPrice
 // sqrtPriceB is the larger of sqrtpCur and the nextPrice
 // calcAmount0Delta = (liquidity * (sqrtPriceB - sqrtPriceA)) / (sqrtPriceB * sqrtPriceA)
-func calcAmount0Delta(liq, sqrtPriceA, sqrtPriceB sdk.Dec) sdk.Dec {
+func calcAmount0Delta(liq, sqrtPriceA, sqrtPriceB sdk.Dec, roundUp bool) sdk.Dec {
```
Comment (mattverse) [inline x/concentrated-liquidity/math.go:36]: What's the reason behind in somecases we round up, some cases we just round down? (Which cases do we round up which cases do we round down and why?)

File: x/concentrated-liquidity/pool.go [hunk 2]
```diff
@@ -100,7 +100,7 @@ func (k Keeper) CalcOutAmtGivenIn(ctx sdk.Context,
 	tokenOutDenom string,
 	swapFee sdk.Dec,
 	priceLimit sdk.Dec,
-	poolId uint64) (tokenIn, tokenOut sdk.Coin, updatedTick sdk.Int, updatedLiquidity sdk.Dec, err error) {
+	poolId uint64) (tokenInDelta, tokenOutDelta sdk.Int, updatedTick sdk.Int, updatedLiquidity sdk.Dec, err error) {
```
Comment (mattverse) [inline x/concentrated-liquidity/pool.go:103]: Im wondering if we should do tokenIn and tokenOut instead? ( I personally think this way would be more useful as API) since the caller would have to recalculate on the actual token In and token Out. What do you think?

File: x/concentrated-liquidity/pool.go [hunk 3]
```diff
@@ -139,15 +139,17 @@ func (k Keeper) CalcOutAmtGivenIn(ctx sdk.Context,
 		liquidity:                p.Liquidity,
 	}
 
-	for swapState.amountSpecifiedRemaining.GT(sdk.ZeroDec()) && !swapState.sqrtPrice.Equal(sqrtPriceLimit) {
+	// had to reintroduce the dust check, we need to figure out a cleaner way to do this
+	for swapState.amountSpecifiedRemaining.GT(sdk.NewDecWithPrec(1, 8)) && !swapState.sqrtPrice.Equal(sqrtPriceLimit) {
```
Comment (mattverse) [inline x/concentrated-liquidity/pool.go:143]: So I personally think we should not reintroduce dust check.
I was digging around with this for loop yesterday, and the closest answer I got to was that it only works with the dust check not because of some sort of minor rounding issues, its because that the calculation being done is off(incorrect) and we're just force exiting from the loop. 

I, myself have to do more investigation on this, and I'm also fine with leaving the dust check for now, but just wanted to let you know that having the dust check means that we still have a point that we missed somewhere!

File: x/concentrated-liquidity/pool_test.go [hunk 4]
```diff
@@ -12,50 +12,182 @@ import (
...
+	swapFee = sdk.ZeroDec()
+	// limit max price impact to 5004 usdc per eth
+	priceLimit = sdk.NewDec(5004)
+
+	// run calculation
+	tokenInDelta, tokenOutDelta, updatedTick, updatedLiquidity, err = s.App.ConcentratedLiquidityKeeper.CalcOutAmtGivenIn(ctx, tokenIn, tokenOutDenom, swapFee, priceLimit, pool.Id)
+	s.Require().NoError(err)
+
+	// we expect to put 41999999 usdc in and in return get .008396 eth back
+	expectedTokenInDelta = sdk.NewInt(41999999)
```
Comment (mattverse) [inline x/concentrated-liquidity/pool_test.go:111]: This doesnt seem like the "delta" but the actual amount of token in?

File: x/concentrated-liquidity/pool_test.go [hunk 5]
```diff
@@ -12,50 +12,182 @@ import (
...
+	// check pool liquidity
+	lowerSqrtPrice, err = s.App.ConcentratedLiquidityKeeper.TickToSqrtPrice(lowerTick)
+	s.Require().NoError(err)
+	upperSqrtPrice, err = s.App.ConcentratedLiquidityKeeper.TickToSqrtPrice(upperTick)
+	s.Require().NoError(err)
+	expectedLiquidity = cl.GetLiquidityFromAmounts(currSqrtPrice, lowerSqrtPrice, upperSqrtPrice, amount0, amount1)
+	s.Require().Equal(expectedLiquidity.String(), updatedLiquidity.String())
+
+	//
+	// TEST 3: two consecutive price ranges
```
Comment (mattverse) [inline x/concentrated-liquidity/pool_test.go:130]: Consider separating these tests out

File: x/concentrated-liquidity/pool_test.go (hunk 4 above)
Comment (czarcas7ic) [reply x/concentrated-liquidity/pool_test.go:111]: It is the delta. Token in delta is what the pool supply changes by, and the token out is also what the pool supply changes by (what you get out), but we take the abs value of that which is why they are both positive

File: x/concentrated-liquidity/pool.go (hunk 3 above)
Comment (czarcas7ic) [reply x/concentrated-liquidity/pool.go]: Agreed this is not something we want long term, but my vs code was having a seizure every time I accidentally reintroduced an inf loop so wanted something to prevent this. Maybe we should introduce a max 100 loop counter for now, wdyt?

File: x/concentrated-liquidity/pool.go (hunk 2 above)
Comment (czarcas7ic) [reply x/concentrated-liquidity/pool.go:103]: These terms are actually sysnonymous, since the delta is what we are putting into and taking out of the pool. But agreed that tokenIn and tokenOut are just simpler to understand / better API

File: x/concentrated-liquidity/pool_test.go (hunk 5 above)
Comment (czarcas7ic) [reply x/concentrated-liquidity/pool_test.go:130]: can you explain what you mean by separating the tests out?

File: x/concentrated-liquidity/pool.go (hunk 2 above)
Comment (czarcas7ic) [reply x/concentrated-liquidity/pool.go]: Changed!

File: x/concentrated-liquidity/math.go (hunk 1 above)
Comment (czarcas7ic) [reply x/concentrated-liquidity/math.go:36]: Discussed offline

File: x/concentrated-liquidity/pool_test.go (hunk 4 above)
Comment (czarcas7ic) [reply x/concentrated-liquidity/pool_test.go]: Changed the name

File: x/concentrated-liquidity/pool_test.go (hunk 5 above)
Comment (mattverse) [reply x/concentrated-liquidity/pool_test.go:130]: Nm, I will create a PR for making this TDD

File: x/concentrated-liquidity/pool.go [hunk 6]
```diff
@@ -100,7 +100,7 @@ func (k Keeper) CalcOutAmtGivenIn(ctx sdk.Context,
 	tokenOutDenom string,
 	swapFee sdk.Dec,
 	priceLimit sdk.Dec,
-	poolId uint64) (tokenIn, tokenOut sdk.Coin, updatedTick sdk.Int, updatedLiquidity sdk.Dec, err error) {
+	poolId uint64) (tokenIn, tokenOut sdk.Int, updatedTick sdk.Int, updatedLiquidity sdk.Dec, err error) {
```
Comment (mattverse) [inline x/concentrated-liquidity/pool.go:103]: Extension from my previous Q: What's the reason for the change to sdk.Int?

IMO it should be sdk.Coin considering that existing CalcOutGiven in returns a coin and the point that the we would be expecting or constructing a coin any way in the caller of this method

File: x/concentrated-liquidity/pool.go [hunk 7]
```diff
@@ -139,15 +139,17 @@ func (k Keeper) CalcOutAmtGivenIn(ctx sdk.Context,
 		liquidity:                p.Liquidity,
 	}
 
+	// had to reintroduce the dust check, we need to figure out a cleaner way to do this
```
Comment (mattverse) [inline x/concentrated-liquidity/pool.go:142]:
```suggestion
```

File: x/concentrated-liquidity/pool.go [hunk 8]
```diff
@@ -164,42 +166,43 @@ func (k Keeper) CalcOutAmtGivenIn(ctx sdk.Context,
...
 			zeroForOne,
 		)
+		swapState.sqrtPrice = sqrtPrice
 
 		swapState.amountSpecifiedRemaining = swapState.amountSpecifiedRemaining.Sub(amountIn)
 		swapState.amountCalculated = swapState.amountCalculated.Add(amountOut)
 
 		// if we have moved to the next tick,
-		if nextSqrtPrice.Equal(sqrtPrice) {
+		if nextSqrtPrice.Equal(swapState.sqrtPrice) {
```
Comment (mattverse) [inline x/concentrated-liquidity/pool.go:175]: I think this can stay the same? (use sqrtPrice)

File: x/concentrated-liquidity/pool.go [hunk 9]
```diff
@@ -164,42 +166,43 @@ func (k Keeper) CalcOutAmtGivenIn(ctx sdk.Context,
...
 	if zeroForOne {
-		tokenIn = sdk.NewCoin(tokenInMin.Denom, amt0)
-		tokenOut = sdk.NewCoin(tokenOutDenom, amt1)
+		tokenIn = (swapState.amountCalculated).TruncateInt()
+		tokenOut = tokenAmountInAfterFee.Add(swapState.amountSpecifiedRemaining).TruncateInt()
 	} else {
-		tokenIn = sdk.NewCoin(tokenInMin.Denom, amt1)
-		tokenOut = sdk.NewCoin(tokenOutDenom, amt0)
+		tokenIn = tokenAmountInAfterFee.Add(swapState.amountSpecifiedRemaining).TruncateInt()
+		tokenOut = swapState.amountCalculated.TruncateInt()
```
Comment (mattverse) [inline x/concentrated-liquidity/pool.go:205]: consider de-duplicating these logic by using a variable above

**Review by mattverse — COMMENTED** [review]
Overall Main logic LGTM, I just have final round of reivews I wanted to adress / ask before merge

File: x/concentrated-liquidity/pool.go (hunk 8 above)
Comment (czarcas7ic) [reply x/concentrated-liquidity/pool.go:175]: they are synonymous now since we we now set swapState.sqrtPrice = sqrtPrice but I will change it back!

File: x/concentrated-liquidity/pool.go (hunk 6 above)
Comment (czarcas7ic) [reply x/concentrated-liquidity/pool.go:103]: Yeah I think I agree with you, thinking back to why I changed this in the first place I think it was for troubleshooting purposes. Will change back

File: x/concentrated-liquidity/math.go [hunk 10]
```diff
@@ -33,48 +33,59 @@ func liquidity1(amount sdk.Int, sqrtPriceA, sqrtPriceB sdk.Dec) sdk.Dec {
...
+		amountIn = calcAmount1Delta(liquidity, sqrtPriceTarget, sqrtPriceCurrent, true)
 	}
-
 	if amountRemaining.GTE(amountIn) {
 		sqrtPriceNext = sqrtPriceTarget
 	} else {
 		sqrtPriceNext = getNextSqrtPriceFromInput(sqrtPriceCurrent, liquidity, amountRemaining, zeroForOne)
 	}
 
+	max := sqrtPriceNext.Equal(sqrtPriceTarget)
```
Comment (p0mvn) [inline x/concentrated-liquidity/math.go:77]: Please add comments about this

File: x/concentrated-liquidity/math.go [hunk 11]
```diff
@@ -33,48 +33,59 @@ func liquidity1(amount sdk.Int, sqrtPriceA, sqrtPriceB sdk.Dec) sdk.Dec {
...
+
 	if zeroForOne {
```
Comment (p0mvn) [inline x/concentrated-liquidity/math.go:79]: I suggest exploring the option of using separate methods for `zero` and `one`.

Also, exploring whether using a design pattern like strategy makes sense here

File: x/concentrated-liquidity/pool.go [hunk 12]
```diff
@@ -187,19 +191,19 @@ func (k Keeper) CalcOutAmtGivenIn(ctx sdk.Context,
...
-		} else {
+		} else if !sqrtPriceStart.Equal(sqrtPrice) {
 			swapState.tick = priceToTick(sqrtPrice.Power(2))
 		}
 	}
 
-	amt0 := tokenAmountInAfterFee.Add(swapState.amountSpecifiedRemaining.Abs()).TruncateInt()
-	amt1 := swapState.amountCalculated.TruncateInt()
+	amt0 := tokenAmountInAfterFee.Add(swapState.amountSpecifiedRemaining).TruncateInt()
+	amt1 := (swapState.amountCalculated).TruncateInt()
```
Comment (p0mvn) [inline x/concentrated-liquidity/pool.go:200]:
```suggestion
	amt1 := swapState.amountCalculated.TruncateInt()
```

Comment (p0mvn) [inline x/concentrated-liquidity/pool.go:200]: Can we add comments to explain truncations here?

File: x/concentrated-liquidity/pool_test.go [hunk 13]
```diff
@@ -12,50 +12,182 @@ import (
 func (s *KeeperTestSuite) TestCalcOutAmtGivenIn() {
 	ctx := s.Ctx
 
+	//
```
Comment (p0mvn) [inline x/concentrated-liquidity/pool_test.go:15]: Let's convert to table-driven please

File: x/concentrated-liquidity/tick.go [hunk 14]
```diff
@@ -87,7 +87,7 @@ func (k Keeper) NextInitializedTick(ctx sdk.Context, poolId uint64, tickIndex in
 	}
 
 	var iter db.Iterator
-	if zeroForOne {
+	if !zeroForOne {
```
Comment (p0mvn) [inline x/concentrated-liquidity/tick.go:90]: I think we should also have separate methods here or a strategy design pattern instead of checking `zeroForOne` everywhere

**Review by p0mvn — CHANGES_REQUESTED** [review]

File: x/concentrated-liquidity/tick.go (hunk 14 above)
Comment (p0mvn) [reply x/concentrated-liquidity/tick.go:90]: Reason: https://softwareengineering.stackexchange.com/questions/147977/is-it-wrong-to-use-a-boolean-parameter-to-determine-behavior

File: x/concentrated-liquidity/lp.go [hunk 15]
```diff
@@ -64,15 +64,15 @@ func (k Keeper) createPosition(ctx sdk.Context, poolId uint64, owner sdk.AccAddr
 
 	// update pool state as well only if we're injecting liquidity at the current tick of the pool
 	if pool.CurrentTick.LT(sdk.NewInt(lowerTick)) {
-		amtDenom0 = calcAmount0Delta(liquidity, sqrtRatioLowerTick, sqrtRatioUpperTick).RoundInt()
+		amtDenom0 = calcAmount0Delta(liquidity, sqrtRatioLowerTick, sqrtRatioUpperTick, false).RoundInt()
```
[the remark this answers is not in the document]
Comment (czarcas7ic) [reply x/concentrated-liquidity/lp.go:67]: As discussed offline, am going to revert the rounding until we can create a test case that actually causes an overflow

File: x/concentrated-liquidity/math.go (hunk 11 above)
Comment (czarcas7ic) [reply x/concentrated-liquidity/math.go:79]: was just gas efficiency, removing now

File: x/concentrated-liquidity/lp.go (hunk 15 above)
[the remark this answers is not in the document]
Comment (czarcas7ic) [reply x/concentrated-liquidity/lp.go:67]: Further discussion offline, rounding both protects from inf loop as well as rounding up on amountIn to prevent removing more from the pool than expected.

File: x/concentrated-liquidity/lp.go [hunk 16]
```diff
@@ -62,17 +62,25 @@ func (k Keeper) createPosition(ctx sdk.Context, poolId uint64, owner sdk.AccAddr
...
+		// we also update the pool liquidity since the virtual liquidity is modified by this position's creation
 	} else if pool.CurrentTick.LT(sdk.NewInt(upperTick)) {
-		amtDenom0 = calcAmount0Delta(liquidity, currentSqrtPrice, sqrtRatioUpperTick).RoundInt()
-		amtDenom1 = calcAmount1Delta(liquidity, currentSqrtPrice, sqrtRatioLowerTick).RoundInt()
+		amtDenom0 = calcAmount0Delta(liquidity, currentSqrtPrice, sqrtRatioUpperTick, false).RoundInt()
+		amtDenom1 = calcAmount1Delta(liquidity, currentSqrtPrice, sqrtRatioLowerTick, false).RoundInt()
 		pool.Liquidity = pool.Liquidity.Add(liquidity)
+
+		// outcome three: position is above current price
+		// tis means position is solely made up of asset1
```
Comment (p0mvn) [inline x/concentrated-liquidity/lp.go:80]:
```suggestion
		// this means position is solely made up of asset1
```

Comment (p0mvn) [inline x/concentrated-liquidity/lp.go:80]: nit: please consider moving comments inside their branches. Now it might look like this is referring to outcome two branch

File: x/concentrated-liquidity/math.go [hunk 17]
```diff
@@ -33,48 +33,59 @@ func liquidity1(amount sdk.Int, sqrtPriceA, sqrtPriceB sdk.Dec) sdk.Dec {
...
-func calcAmount0Delta(liq, sqrtPriceA, sqrtPriceB sdk.Dec) sdk.Dec {
+func calcAmount0Delta(liq, sqrtPriceA, sqrtPriceB sdk.Dec, roundUp bool) sdk.Dec {
 	if sqrtPriceA.GT(sqrtPriceB) {
 		sqrtPriceA, sqrtPriceB = sqrtPriceB, sqrtPriceA
 	}
 	diff := sqrtPriceB.Sub(sqrtPriceA)
 	mult := liq
+	// if calculating for amountIn, we round up
+	// if calculating for amountOut, we don't round
+	// this is to prevent removing more from the pool than expected due to rounding
```
Comment (p0mvn) [inline x/concentrated-liquidity/math.go:44]: Could you elaborate a bit on this please? Why would we remove more than expected from pool? Where exactly the difference leads? For example, offline we discussed an infinite loop. Where is that loop?

File: x/concentrated-liquidity/math.go [hunk 18]
```diff
@@ -33,48 +33,59 @@ func liquidity1(amount sdk.Int, sqrtPriceA, sqrtPriceB sdk.Dec) sdk.Dec {
...
 // calcAmount1Delta = liq * (sqrtPriceB - sqrtPriceA)
-func calcAmount1Delta(liq, sqrtPriceA, sqrtPriceB sdk.Dec) sdk.Dec {
+func calcAmount1Delta(liq, sqrtPriceA, sqrtPriceB sdk.Dec, roundUp bool) sdk.Dec {
 	if sqrtPriceA.GT(sqrtPriceB) {
 		sqrtPriceA, sqrtPriceB = sqrtPriceB, sqrtPriceA
 	}
 	diff := sqrtPriceB.Sub(sqrtPriceA)
+	// if calculating for amountIn, we round up
+	// if calculating for amountOut, we don't round
+	// this is to prevent removing more from the pool than expected due to rounding
```
Comment (p0mvn) [inline x/concentrated-liquidity/math.go:62]: ditto to expand on rounding

File: x/concentrated-liquidity/math.go [hunk 19]
```diff
@@ -33,48 +33,59 @@ func liquidity1(amount sdk.Int, sqrtPriceA, sqrtPriceB sdk.Dec) sdk.Dec {
...
 		sqrtPriceNext = sqrtPriceTarget
 	} else {
 		sqrtPriceNext = getNextSqrtPriceFromInput(sqrtPriceCurrent, liquidity, amountRemaining, zeroForOne)
 	}
 
 	if zeroForOne {
-		amountIn = calcAmount0Delta(liquidity, sqrtPriceNext, sqrtPriceCurrent)
-		amountOut = calcAmount1Delta(liquidity, sqrtPriceNext, sqrtPriceCurrent)
+		amountIn = calcAmount0Delta(liquidity, sqrtPriceNext, sqrtPriceCurrent, true)
+		amountOut = calcAmount1Delta(liquidity, sqrtPriceNext, sqrtPriceCurrent, false)
```
Comment (p0mvn) [inline x/concentrated-liquidity/math.go:85]: Why do we round up `amountIn` calculation here but round down `amountOut`? I think this deserves a comment

File: x/concentrated-liquidity/pool_test.go (hunk 13 above)
Comment (czarcas7ic) [reply x/concentrated-liquidity/pool_test.go]: Just merged Matts branch that did this!

File: x/concentrated-liquidity/math.go (hunk 17 above)
Comment (czarcas7ic) [reply x/concentrated-liquidity/math.go:44]: elaborated on this further. this should also address your concern in math.go. If it doesn't, I don't have any other way I can further explain this so will have to get on a call to further hash out what we can say to make it clearer.

File: x/concentrated-liquidity/math.go [hunk 20]
```diff
@@ -33,48 +33,67 @@ func liquidity1(amount sdk.Int, sqrtPriceA, sqrtPriceB sdk.Dec) sdk.Dec {
...
 	if sqrtPriceA.GT(sqrtPriceB) {
 		sqrtPriceA, sqrtPriceB = sqrtPriceB, sqrtPriceA
 	}
 	diff := sqrtPriceB.Sub(sqrtPriceA)
 	mult := liq
+	// if calculating for amountIn, we round up
+	// if calculating for amountOut, we don't round at all
+	// this is to prevent removing more from the pool than expected due to rounding
+	// example: we calculate 1000000.9999999 uusdc (~$1) amountIn and 2000000.999999 uosmo amountOut
+	// we would want the used to put in 1000001 uusdc rather than 1000000 uusdc to ensure we are charging enough for the amount they are removing
```
Comment (p0mvn) [inline x/concentrated-liquidity/math.go:46]:
```suggestion
	// we would want the user to put in 1000001 uusdc rather than 1000000 uusdc to ensure we are charging enough for the amount they are removing
```

File: x/concentrated-liquidity/math.go (hunk 17 above)
Comment (p0mvn) [reply x/concentrated-liquidity/math.go:44]: This is perfect, thank you

**Review by p0mvn — APPROVED** [review]
Great work! Comments are super helpful.

I focused only on core logic and only skimmed the tests. Planning to review the tests later outside of the PR.

Also, as discussed offline, it would be great to investigate separately whether this approach would reduce the complexity from passing in `zeroForOne` booleans 
https://github.com/osmosis-labs/osmosis/issues/3218
